book/src/proposals/booting-from-snapshot.md

```diff
@@ -12,20 +12,102 @@ The validator may be fed a snapshot and gossip info for a different (though not
+
+Define the "safety" condition to be: 
+
+A validator can run some procedure to determine whether it can vote on `T` without violating the lockouts of any vote in `L`. The procedure for this is:
+
+1) If `T` < last locktower vote, return false,
+2) Apply `T` to locktower state `L`. Pope off all votes `S_i` where `!is_locked(S_i, T)`. `T` must now be the top of the tower.
+3) If for any remaining vote `L_i` in `L` `!is_ancestor(L_i, T)`, return false.
+4) Otherwise, return true.
+
+# Achieving Safety
+
+Define the "Safety Criteria" to be: Given any descendant of `S`, `S_d`, and any slot `L_i` in locktower, the validator is able to determine `is_ancestor(L_i, S_d)`.
+
+Assume the "Safety Criteria" is true, we show we can then achieve the "safety" condition:
+
+Proof: 
+A validator wants to determine whether it can vote on `T`. `T` must be a descendant of `S` because the validator does not play any state for non-descendants of `S` when booting from the snapshot. From assuming "Safety Criteria" above, this means for any `L_i` we can determine `is_ancestor(L_i, T)`. This means we have all the tools to run the algorithm from the safety definition.
+
+Thus to achieve safety, we want to design the snapshotting system such that the "Safety Criteria" is met.
+
+# Implementing "Safety Criteria"
+
+In our implementation, we assume that at the time the validator starts from the snapshot, `L_i + N >= S` for some large `N`. 
+
+Design: 
+
+1) The snapshot is augmented to store the last `N` ancestors of `S`. These ancestors are  incorporated into the bank hash so they can be verified when a validator unpacks a snapshot. Call this set `N_Ancestors`.
+
+2) We reject any snapshots `S` where `S` is less than the root of `L` and `S` is not in the list of roots in blocktree because this means `S` is for a different fork than the one this validiator last rooted. Thus it's critical for consistency in `replay_stage` that the order of events when setting a new root is:
+    a) Write the root to locktower
+    b) Write the root to blocktree
+    c) Generate snapshot for that root
+
+3) On startup, make sure the current root in locktower exists in blocktree, if not (there was a crash between 2a and 2b), then rewrite the root to blocktree.
+
+4) On startup, the validator boots from the snapshot `S`, then replays all descendant blocks of `S` that exist in this validator's ledger, building banks which are then stored in an output `BankForks`. This is done in `blocktree_processor.rs`. The root of this `BankForks` is set to `S`.
+
+
+We now implement the "Safety Criteria" in cases:
+
+Case 1: Calculating `is_ancestor(L_i, S_d)` when `L_i` < `S`:
+
+Search the list of `N_Ancestors` ancestors of `S` to check if `L_i` is an ancestor of `S`.
+
+Proof of Correctness for Case 1:
+This case is equivalent to determining `is_ancestor(L_i, S)`. Because `L_i >= S + N`, and `N_Ancestors` has length `N`, then if `L_i`,is an ancestor, it has to be a member of `N_Ancestors`. 
```

Comment:
Seems like this might be might be easier/clearer to explain with the transitive property: `is_ancestor(L_i, S_d) <=> ( is_ancestor(L_i, S) && is_ancestor(S, S_d) )` where `is_ancestor(S, S_d)` is true by the definition of a snapshot.

Reply:
Yeah, thats clearer!